File: TerraObserver/TO.Domains/Functions/Maps/RealEarthLandGenerator.fs
namespace TO.Domains.Functions.Maps

open Godot
open TO.Domains.Functions.HexGridCoords
open TO.Domains.Functions.HexSpheres.Components
open TO.Domains.Functions.HexSpheres.Components.Tiles
open TO.Domains.Types.Configs
open TO.Domains.Types.Friflos
open TO.Domains.Types.HexSpheres.Components
open TO.Domains.Types.HexSpheres.Components.Tiles
open TO.Domains.Types.Maps

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-07-05 11:41:05
module RealEarthLandGeneratorCommand =
    let createLand
        (env:
            'E
                when 'E :> IHexMapGeneratorQuery
                and 'E :> IPlanetConfigQuery
                and 'E :> IEntityStoreCommand
                and 'E :> ITileQuery)
        (landGen: IRealEarthLandGenerator)
        =
        let water = env.HexMapGenerator.DefaultWaterLevel
        let elevationStep = env.PlanetConfig.ElevationStep
        let mutable landCount = 0
        let worldMap = landGen.WorldMap.GetImage()

        env.ExecuteInCommandBuffer(fun cb ->
            for tile in env.GetAllTiles() do
                let mutable tileValue = tile |> Tile.value

                let lonLat =
                    env.GetSphereAxial tile |> SphereAxial.toLonLat |> LonLatCoords.toVector2

                let percentX = Mathf.Remap(lonLat.X, 180f, -180f, 0f, 1f) // 西经为正，所以这里得反一下
                let percentY = Mathf.Remap(lonLat.Y, 90f, -90f, 0f, 1f) // 北纬为正，所以这里得反一下
                let mutable x = int (4096f * percentX) // 宽度 4096

                if x >= 4096 then
                    x <- 4095 // 不知道为啥 Mathf.Clamp 限制不了…… 手动限制一下

                let mutable y = int (2048f * percentY) // 高度 2048

                if y >= 2048 then
                    y <- 2047 // 不知道为啥 Mathf.Clamp 限制不了…… 手动限制一下

                let mutable elevation = 0
                let color = worldMap.GetPixel(x, y)

                if color.R > 0.9f then
                    // 陆地
                    elevation <- int (color.G * float32 (elevationStep + 1 - water))

                    if elevation = elevationStep + 1 - water then
                        elevation <- elevation - 1

                    elevation <- elevation + water
                    landCount <- landCount + 1
                elif color.G > 0.1f then
                    // 湖区
                    // BUG: 现在的实现可能出现高于相邻地块的水面
                    elevation <- int (color.G * float32 (elevationStep + 1 - water))

                    if elevation = elevationStep + 1 - water then
                        elevation <- elevation - 1

                    elevation <- elevation + water - 1
                    tileValue <- tileValue |> TileValue.withWaterLevel (elevation + 1)
                else
                    // 海洋
                    elevation <- int (color.B * float32 water)

                    if elevation = water then
                        elevation <- elevation - 1

                tileValue <- tileValue |> TileValue.withElevation elevation
                cb.AddComponent<TileValue>(tile.Id, &tileValue))

        landCount

File: TerraObserver/Scenes/Maps/Models/RealEarthLandGenerator.cs
using Godot;
using TO.Domains.Types.Maps;

namespace TerraObserver.Scenes.Maps.Models;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-07-05 11:38:50
[Tool]
[GlobalClass]
public partial class RealEarthLandGenerator : LandGenerator, IRealEarthLandGenerator
{
    // 其实这里可以直接导入 Image, 在导入界面选择导入类型。但是导入 Image 的场景 tscn 文件会大得吓人……（等于直接按像素写一遍）
    // 这里是否改为 Image 需要权衡，目前 Texture2D.GetImage() 会耗时比较长，但是提前把 Image 加载到内存的话，消耗很大。
    // 所以目前把四张图片压到同一张图片的 RGB 通道里了（陆地掩码（去除湖区） R，地形 G，海洋测深 B），并压缩大小到 4096x2048
    [Export] public Texture2D? WorldMap { get; set; }
}